### PR TITLE
Add back auto deploy

### PR DESCRIPTION
**Pull request by RMartinOscar**
Co-authored By: @notAreYouScared

image

image
image

image
image

Supersedes #592
Needs https://github.com/pelican-dev/wings/pull/46

### CODE REVIEW

**Comment by RMartinOscar** [comment]
Would be cool if we had an env variable like ```ENV APP_DOCKER=true``` in [Dockerfile](https://github.com/pelican-dev/panel/blob/main/Dockerfile) so we can set the default value of `get('docker')`
Like https://github.com/pelican-dev/panel/blob/a9e11cd246493d713e67ac8c182683e61ff16866/app/Filament/Resources/NodeResource/Pages/EditNode.php#L440
```php
->default(env('APP_DOCKER', false))
```

File: app/Services/Nodes/NodeAutoDeployService.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     * Generates a new API key for the logged-in user with only permission to read
+     * nodes, and returns that as the deployment key for a node.
+     *
+     * @throws \App\Exceptions\Model\DataValidationException
+     */
+    public function handle(Request $request, Node $node, ?bool $docker = false): ?string
+    {
+        /** @var ApiKey|null $key */
+        $key = ApiKey::query()
+            ->where('user_id', $request->user()->id)
```
Comment (Boy132) [inline app/Services/Nodes/NodeAutoDeployService.php:30]: Why the user check? For application keys this shouldn't be needed.

**Review by Boy132 — CHANGES_REQUESTED** [review]

File: app/Services/Nodes/NodeAutoDeployService.php (hunk 1 above)
Comment (RMartinOscar) [reply app/Services/Nodes/NodeAutoDeployService.php:30]: It won't find the keys without it thus creating thousand duplicates.

Comment (RMartinOscar) [reply app/Services/Nodes/NodeAutoDeployService.php:30]: Im gonna try again but for some reason it didn't work on my side when i was testing that.

File: app/Services/Nodes/NodeAutoDeployService.php [hunk 2]
```diff
@@ -47,7 +46,7 @@ public function handle(Request $request, Node $node, ?bool $docker = false): ?st
             sprintf(
                 '%s wings configure --panel-url %s --token %s --node %d%s',
                 $docker ? 'docker compose exec -it' : 'sudo',
-                config('app.url'),
+                route('index'),
```
Comment (Boy132) [inline app/Services/Nodes/NodeAutoDeployService.php:49]: Revert that. The `APP_URL` is correct here.

File: app/Models/ApiKey.php [hunk 3]
```diff
@@ -66,15 +66,18 @@ class ApiKey extends Model
 
     /**
      * Maximum number of Api keys that a user can have.
+     *
+     * @deprecated
```
Comment (Boy132) [inline app/Models/ApiKey.php:70]: You can remove the consts instead of deprecating them.

**Review by Boy132 — DISMISSED** [review]

File: app/Filament/Resources/ApiKeyResource/Pages/CreateApiKey.php [hunk 4]
```diff
@@ -23,7 +23,7 @@ public function form(Form $form): Form
         return $form
             ->schema([
                 Hidden::make('identifier')->default(ApiKey::generateTokenIdentifier(ApiKey::TYPE_APPLICATION)),
-                Hidden::make('token')->default(str_random(ApiKey::KEY_LENGTH)),
+                Hidden::make('token')->default(str_random(config('api.key.secret_length', 32))),
```
Comment (lancepioch) [inline app/Filament/Resources/ApiKeyResource/Pages/CreateApiKey.php:26]: This should NOT be configurable!

File: app/Http/Controllers/Api/Client/ApiKeyController.php [hunk 5]
```diff
@@ -29,7 +29,7 @@ public function index(ClientApiRequest $request): array
      */
     public function store(StoreApiKeyRequest $request): array
     {
-        if ($request->user()->apiKeys->count() >= 25) {
+        if ($request->user()->apiKeys->count() >= config('api.key.limit', 25)) {
```
Comment (lancepioch) [inline app/Http/Controllers/Api/Client/ApiKeyController.php:32]: Use the constant from before

File: app/Services/Nodes/NodeAutoDeployService.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+     * nodes, and returns that as the deployment key for a node.
+     *
+     * @throws \App\Exceptions\Model\DataValidationException
+     */
+    public function handle(Request $request, Node $node, ?bool $docker = false): ?string
+    {
+        /** @var ApiKey|null $key */
+        $key = ApiKey::query()
+            ->where('key_type', ApiKey::TYPE_APPLICATION)
+            ->where('r_nodes', 1)
```
Comment (lancepioch) [inline app/Services/Nodes/NodeAutoDeployService.php:31]:
```suggestion
            ->where('r_nodes', true)
```

File: app/Services/Nodes/NodeAutoDeployService.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            ->first();
+
+        // We couldn't find a key that exists for this user with only permission for
+        // reading nodes. Go ahead and create it now.
+        if (!$key) {
+            $key = $this->keyCreationService->setKeyType(ApiKey::TYPE_APPLICATION)->handle([
+                'memo' => 'Automatically generated node deployment key.',
+                'user_id' => $request->user()->id,
+            ], ['r_nodes' => 1]);
```
Comment (lancepioch) [inline app/Services/Nodes/NodeAutoDeployService.php:40]:
```suggestion
            ], ['r_nodes' => true]);
```

File: app/Services/Nodes/NodeAutoDeployService.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        $token = $key->identifier . $key->token;
+
+        return $token ?
```
Comment (lancepioch) [inline app/Services/Nodes/NodeAutoDeployService.php:45]: Make it easier to read by getting rid of the ternary

```php
if (! $token) {
    return null;
}
```

File: config/api.php
Comment (lancepioch) [inline config/api.php:1]: I think this will only cause issues for people that are now allowed to modify it.

**Review by lancepioch — CHANGES_REQUESTED** [review]
I would say remove the api config, put the max api keys in the panel configuration file, but the others (identifier_length and secret_length) should remain constants. Otherwise, looks good!

**Review by lancepioch — APPROVED** [review]

**Pull request merged**